File: 1955-count-number-of-special-subsequences/1955-count-number-of-special-subsequences.py
class Solution:
    def countSpecialSubsequences(self, nums: List[int]) -> int:
        MOD = 10**9 + 7
        dp = [1, 0, 0, 0]
        for num in nums:
            idx = num + 1
            dp[idx] += (dp[idx] + dp[idx-1]) 
            dp[idx] = dp[idx] % MOD
        return dp[-1]

File: 1955-count-number-of-special-subsequences/1955-count-number-of-special-subsequences.rs

impl Solution {
    pub fn count_special_subsequences(nums: Vec<i32>) -> i32 {
        let m: i64 = 1_000_000_007;

        let mut a: i64 = 0;
        let mut b: i64 = 0;
        let mut c: i64 = 0;

        for x in nums {
            match x {
                0 => {
                    a = (2 * a + 1) % m;
                }
                1 => {
                    b = (2 * b + a) % m;
                }
                2 => {
                    c = (2 * c + b) % m;
                }
                _ => {}
            }
        }

        c as i32
    }
}

